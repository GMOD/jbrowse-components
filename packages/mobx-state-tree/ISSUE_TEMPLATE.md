**If your issue isn't a bug report, please consider using discussion threads instead of opening an issue: https://github.com/mobxjs/mobx-state-tree/discussions**
