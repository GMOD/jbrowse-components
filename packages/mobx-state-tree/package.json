{
  "name": "@jbrowse/mobx-state-tree",
  "version": "5.4.2",
  "description": "Opinionated, transactional, MobX powered state container",
  "main": "dist/index.js",
  "module": "esm/index.js",
  "sideEffects": false,
  "scripts": {
    "clean": "rimraf dist esm",
    "prebuild": "yarn clean",
    "build:esm": "tsup  --format esm src/index.ts --dts -d esm",
    "build:es5": "tsup  src/index.ts --dts",
    "build": "yarn build:es5 && yarn build:esm",
    "typecheck": "tsc --noEmit"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/mobxjs/@jbrowse/@jbrowse/mobx-state-tree.git"
  },
  "author": "Michel Weststrate",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mobxjs/mobx-state-tree/issues"
  },
  "files": [
    "dist/"
  ],
  "peerDependencies": {
    "mobx": "^6.3.0"
  },
  "keywords": [
    "mobx",
    "mobx-state-tree",
    "promise",
    "reactive",
    "frp",
    "functional-reactive-programming",
    "state management"
  ],
  "dependencies": {
    "ts-essentials": "^10.0.4"
  },
  "devDependencies": {
    "@swc/core": "^1.11.20",
    "tsup": "^8.4.0"
  }
}
