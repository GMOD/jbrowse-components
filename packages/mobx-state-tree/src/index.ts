// tslint:disable-next-line:no_unused-variable
import { IObservableArray, ObservableMap } from 'mobx'

/* all code is initially loaded through internal, to avoid circular dep issues */
export {
  typecheck,
  escapeJsonPath,
  unescapeJsonPath,
  joinJsonPath,
  splitJsonPath,
  decorate,
  addMiddleware,
  process,
  isStateTreeNode,
  flow,
  castFlowReturn,
  applyAction,
  onAction,
  recordActions,
  createActionTrackingMiddleware,
  createActionTrackingMiddleware2,
  setLivelinessChecking,
  getLivelinessChecking,
  setLivelynessChecking, // to be deprecated
  getType,
  getChildType,
  onPatch,
  onSnapshot,
  applyPatch,
  recordPatches,
  protect,
  unprotect,
  isProtected,
  applySnapshot,
  getSnapshot,
  hasParent,
  getParent,
  hasParentOfType,
  getParentOfType,
  getRoot,
  getPath,
  getPathParts,
  isRoot,
  resolvePath,
  resolveIdentifier,
  getIdentifier,
  tryResolve,
  getRelativePath,
  clone,
  detach,
  destroy,
  isAlive,
  addDisposer,
  getEnv,
  walk,
  getMembers,
  getPropertyMembers,
  cast,
  castToSnapshot,
  castToReferenceSnapshot,
  isType,
  isArrayType,
  isFrozenType,
  isIdentifierType,
  isLateType,
  isLiteralType,
  isMapType,
  isModelType,
  isOptionalType,
  isPrimitiveType,
  isReferenceType,
  isRefinementType,
  isUnionType,
  tryReference,
  isValidReference,
  getNodeId,
  getRunningActionContext,
  isActionContextChildOf,
  isActionContextThisOrChildOf,
  types,
  types as t, // We do this as a less ambiguous term for the traditional types module, which sometimes gets confused when discussing "types" in general
  toGeneratorFunction,
  toGenerator,
} from './internal'

export type {
  IModelType,
  IAnyModelType,
  IDisposer,
  IMSTMap,
  IMapType,
  IMSTArray,
  IArrayType,
  IType,
  IAnyType,
  ModelPrimitive,
  ISimpleType,
  IComplexType,
  IAnyComplexType,
  IReferenceType,
  IStateTreeNode,
  IAnyStateTreeNode,
  IActionRecorder,
  ISerializedActionCall,
  ReferenceIdentifier,
  ISnapshotProcessor,
  ISnapshotProcessors,
  OnReferenceInvalidated,
  OnReferenceInvalidatedEvent,
  ReferenceOptions,
  ReferenceOptionsGetSet,
  ReferenceOptionsOnInvalidated,
  IActionContext,
  TypeOfValue,
  IModelReflectionData,
  IModelReflectionPropertiesData,
  IMaybeIType,
  IMaybe,
  IMaybeNull,
  IOptionalIType,
  OptionalDefaultValueOrFunction,
  ValidOptionalValue,
  ValidOptionalValues,
  LivelynessMode, // to be deprecated
  ModelSnapshotType,
  ModelCreationType,
  ModelSnapshotType2,
  ModelCreationType2,
  ModelInstanceType,
  ModelInstanceTypeProps,
  ModelPropertiesDeclarationToProperties,
  ModelProperties,
  ModelPropertiesDeclaration,
  ModelActions,
  ITypeUnion,
  CustomTypeOptions,
  UnionOptions,
  Instance,
  SnapshotIn,
  SnapshotOut,
  SnapshotOrInstance,
  TypeOrStateTreeNodeToStateTreeNode,
  UnionStringArray,
  IMiddlewareEvent,
  IActionTrackingMiddleware2Call,
  IMiddlewareHandler,
  IMiddlewareEventType,
  IActionTrackingMiddlewareHooks,
  IActionTrackingMiddleware2Hooks,
  IJsonPatch,
  IReversibleJsonPatch,
  _CustomCSProcessor,
  _CustomOrOther,
  _CustomJoin,
  _NotCustomized,
  LivelinessMode,
  IPatchRecorder,
} from './internal'
