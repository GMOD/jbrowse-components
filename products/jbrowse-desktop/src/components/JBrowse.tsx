import React from 'react'

// jbrowse
import { App } from '@jbrowse/app-core'
import { CssBaseline } from '@mui/material'
import { ThemeProvider } from '@mui/material/styles'
import { observer } from 'mobx-react'

// locals
import type { DesktopRootModel } from '../rootModel'
import type PluginManager from '@jbrowse/core/PluginManager'

const JBrowseNonNullRoot = observer(function ({
  rootModel,
}: {
  rootModel: DesktopRootModel
}) {
  const { session, error } = rootModel

  if (error) {
    // eslint-disable-next-line @typescript-eslint/only-throw-error
    throw error
  }

  return session ? (
    <ThemeProvider theme={session.theme}>
      <CssBaseline />
      <App session={session} />
    </ThemeProvider>
  ) : null
})

const JBrowse = observer(function ({
  pluginManager,
}: {
  pluginManager: PluginManager
}) {
  const { rootModel } = pluginManager
  return rootModel ? (
    <JBrowseNonNullRoot rootModel={rootModel as DesktopRootModel} />
  ) : null
})

export default JBrowse
