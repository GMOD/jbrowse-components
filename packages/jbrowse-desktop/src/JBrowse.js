import '@gmod/jbrowse-core/fonts/material-icons.css'
import { useDebounce } from '@gmod/jbrowse-core/util'
import {
  App,
  FatalErrorDialog,
  StartScreen,
  theme,
} from '@gmod/jbrowse-core/ui'

import CircularProgress from '@material-ui/core/CircularProgress'
import CssBaseline from '@material-ui/core/CssBaseline'

import { ThemeProvider } from '@material-ui/styles'

import { observer } from 'mobx-react'
import { onSnapshot } from 'mobx-state-tree'
import ErrorBoundary from 'react-error-boundary'
import React, { useEffect, useState } from 'react'
import rootModel from './rootModel'
import 'typeface-roboto'

const debounceMs = 1000

function useJBrowseDesktop() {
  const { electron = {} } = window
  const { desktopCapturer, ipcRenderer } = electron
  const [loaded, setLoaded] = useState(false)
  const [root, setRoot] = useState({})
  const [firstLoad, setFirstLoad] = useState(true)
  const [sessionSnapshot, setSessionSnapshot] = useState()
  const [configSnapshot, setConfigSnapshot] = useState()
  const debouncedSessionSnapshot = useDebounce(sessionSnapshot, debounceMs)
  const debouncedConfigSnapshot = useDebounce(configSnapshot, debounceMs)

  const { session, jbrowse } = root
  if (firstLoad && session) setFirstLoad(false)

  useEffect(() => {
    async function loadConfig() {
      try {
        setRoot(
          rootModel.create({
            jbrowse: Object.assign(await ipcRenderer.invoke('loadConfig'), {
              configuration: {
                rpc: {
                  defaultDriver: 'ElectronRpcDriver',
                },
              },
            }),
          }),
        )
        setLoaded(true)
      } catch (e) {
        setLoaded(() => {
          // throw to error boundary
          throw e
        })
      }
    }

    loadConfig()
  }, [ipcRenderer])

  useEffect(() => {
    return jbrowse
      ? onSnapshot(jbrowse, snap => {
          setConfigSnapshot(snap)
        })
      : () => {}
  }, [jbrowse])
  useEffect(() => {
    return session
      ? onSnapshot(session, snap => {
          setSessionSnapshot(snap)
        })
      : () => {}
  }, [session])

  useEffect(() => {
    ;(async () => {
      if (debouncedSessionSnapshot) {
        const sources = await desktopCapturer.getSources({
          types: ['window'],
          thumbnailSize: { width: 500, height: 500 },
        })
        const jbWindow = sources.find(source => source.name === 'JBrowse')
        const screenshot = jbWindow.thumbnail.toDataURL()
        ipcRenderer.send('saveSession', debouncedSessionSnapshot, screenshot)
      }
    })()
  }, [debouncedSessionSnapshot, desktopCapturer, ipcRenderer])

  useEffect(() => {
    if (debouncedConfigSnapshot) {
      ipcRenderer.send('saveConfig', debouncedConfigSnapshot)
    }
  }, [debouncedConfigSnapshot, ipcRenderer])
  return [root, loaded, firstLoad]
}

const JBrowse = observer(() => {
  const [root, loaded, firstLoad] = useJBrowseDesktop()
  const debouncedLoaded = useDebounce(loaded, 400)
  useEffect(() => {
    if (root) {
      window.MODEL = root.session
      window.ROOTMODEL = root
    }
  }, [root, root.session])

  // Use a debounce loaded here to let the circle spinner give a tiny more turn
  // which looks better
  if (debouncedLoaded) {
    return root.session ? (
      <App session={root.session} />
    ) : (
      <StartScreen root={root} bypass={firstLoad} />
    )
  }
  return (
    <CircularProgress
      style={{
        position: 'fixed',
        top: '50%',
        left: '50%',
        marginTop: -25,
        marginLeft: -25,
      }}
      size={50}
    />
  )
})

export default props => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <ErrorBoundary FallbackComponent={FatalErrorDialog}>
        <JBrowse {...props} />
      </ErrorBoundary>
    </ThemeProvider>
  )
}
