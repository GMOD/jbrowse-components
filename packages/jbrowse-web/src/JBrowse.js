import { readConfObject } from '@gmod/jbrowse-core/configuration'
import '@gmod/jbrowse-core/fonts/material-icons.css'
import { App, theme, FatalErrorDialog } from '@gmod/jbrowse-core/ui'
import {
  toUrlSafeB64,
  fromUrlSafeB64,
  useDebounce,
  inDevelopment,
  mergeConfigs,
} from '@gmod/jbrowse-core/util'
import ErrorBoundary from 'react-error-boundary'

import { openLocation } from '@gmod/jbrowse-core/util/io'

// material-ui
import CircularProgress from '@material-ui/core/CircularProgress'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/styles'

import { observer } from 'mobx-react'
import { onSnapshot } from 'mobx-state-tree'
import { UndoManager } from 'mst-middlewares'
import React, { useEffect, useState } from 'react'
import 'typeface-roboto'

import JBrowseRootModel from './rootModel'

async function parseConfig(configLoc) {
  const config = JSON.parse(await openLocation(configLoc).readFile('utf8'))
  if (inDevelopment) {
    config.datasets = mergeConfigs(
      config,
      JSON.parse(
        await openLocation({ uri: 'test_data/config_in_dev.json' }).readFile(
          'utf8',
        ),
      ),
    )
  }
  return config
}

function useJBrowseWeb(config, initialState) {
  const [loaded, setLoaded] = useState(false)
  const [rootModel, setRootModel] = useState(initialState || {})
  const [urlSnapshot, setUrlSnapshot] = useState()
  const [configSnapshot, setConfigSnapshot] = useState(config || {})
  const debouncedUrlSnapshot = useDebounce(urlSnapshot, 400)

  const { session, jbrowse } = rootModel || {}
  const useLocalStorage = jbrowse
    ? readConfObject(jbrowse.configuration, 'useLocalStorage')
    : false

  const useUpdateUrl = jbrowse
    ? readConfObject(jbrowse.configuration, 'updateUrl')
    : false

  // This serializes the session to URL
  useEffect(() => {
    if (debouncedUrlSnapshot) {
      const { origin, pathname } = document.location
      window.history.replaceState(
        {},
        '',
        `${origin}${pathname}?session=${toUrlSafeB64(
          JSON.stringify(debouncedUrlSnapshot),
        )}`,
      )
    }
  }, [debouncedUrlSnapshot])

  // This updates savedSession list on the rootModel
  useEffect(() => {
    if (rootModel && rootModel.session && debouncedUrlSnapshot) {
      rootModel.jbrowse.updateSavedSession(debouncedUrlSnapshot)
    }
  }, [debouncedUrlSnapshot, rootModel])
  useEffect(() => {
    try {
      setRootModel(JBrowseRootModel.create({ jbrowse: configSnapshot }))
    } catch (error) {
      // if it failed to load, it's probably a problem with the saved sessions,
      // so just delete them and try again
      setRootModel(
        JBrowseRootModel.create({
          jbrowse: { ...configSnapshot, savedSessions: [] },
        }),
      )
    }
  }, [configSnapshot])

  // This loads a config from localStorage or a configSnapshot or a config.json file
  useEffect(() => {
    ;(async () => {
      try {
        if (initialState) {
          setRootModel(initialState)
        } else {
          const localStorageConfig =
            useLocalStorage && localStorage.getItem('jbrowse-web-data')

          if (localStorageConfig) {
            setConfigSnapshot(JSON.parse(localStorageConfig))
          }
          if (configSnapshot.uri || configSnapshot.localPath) {
            setConfigSnapshot(await parseConfig(config))
          }
        }
      } catch (e) {
        setLoaded(() => {
          // throw to error boundary
          throw e
        })
      }
    })()
  }, [
    config,
    configSnapshot.localPath,
    configSnapshot.uri,
    initialState,
    useLocalStorage,
  ])

  // finalize rootModel and setLoaded
  useEffect(() => {
    const params = new URL(document.location).searchParams
    const urlSession = params.get('session')
    if (rootModel && rootModel.jbrowse) {
      if (urlSession) {
        const savedSessionIndex = rootModel.jbrowse.savedSessionNames.indexOf(
          urlSession,
        )
        if (savedSessionIndex !== -1) {
          rootModel.setSession(
            rootModel.jbrowse.savedSessions[savedSessionIndex],
          )
        } else {
          rootModel.setSession(JSON.parse(fromUrlSafeB64(urlSession)))
        }
      } else {
        const localStorageSession = localStorage.getItem('jbrowse-web-session')
        if (localStorageSession) {
          rootModel.setSession(JSON.parse(localStorageSession))
        }
      }
      if (!rootModel.session) {
        if (rootModel.jbrowse && rootModel.jbrowse.savedSessions.length) {
          const { name } = rootModel.jbrowse.savedSessions[0]
          rootModel.activateSession(name)
        } else {
          rootModel.setDefaultSession()
        }
      }

      rootModel.setHistory(
        UndoManager.create({}, { targetStore: rootModel.session }),
      )
      setLoaded(true)
    }
  }, [rootModel])

  // make some things available globally for testing
  // e.g. window.MODEL.views[0] in devtools
  useEffect(() => {
    if (loaded) {
      window.MODEL = rootModel.session
      window.ROOTMODEL = rootModel
    }
  }, [loaded, rootModel])

  // set session in localstorage
  useEffect(
    () =>
      useLocalStorage && loaded
        ? onSnapshot(rootModel.session, snapshot => {
            localStorage.setItem(
              'jbrowse-web-session',
              JSON.stringify(snapshot),
            )
          })
        : () => {},
    [loaded, rootModel, useLocalStorage],
  )

  // set jbrowse-web data in localstorage
  useEffect(
    () =>
      useLocalStorage && loaded
        ? onSnapshot(rootModel.jbrowse, snapshot => {
            localStorage.setItem('jbrowse-web-data', JSON.stringify(snapshot))
          })
        : () => {},
    [loaded, rootModel, useLocalStorage],
  )

  useEffect(
    () =>
      session && useUpdateUrl
        ? onSnapshot(session, snapshot => {
            setUrlSnapshot(snapshot)
          })
        : () => {},
    [useUpdateUrl, session],
  )

  return [loaded, rootModel]
}

const JBrowse = observer(({ config, initialState }) => {
  const [loaded, root] = useJBrowseWeb(config, initialState)
  const debouncedLoaded = useDebounce(loaded, 400)
  // Use a debounce loaded here to let the circle spinner give a tiny more turn
  // which looks better
  return !debouncedLoaded ? (
    <CircularProgress
      style={{
        position: 'fixed',
        top: '50%',
        left: '50%',
        marginTop: -25,
        marginLeft: -25,
      }}
      size={50}
    />
  ) : (
    <App session={root.session} />
  )
})

export default props => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <ErrorBoundary FallbackComponent={FatalErrorDialog}>
        <JBrowse {...props} />
      </ErrorBoundary>
    </ThemeProvider>
  )
}
