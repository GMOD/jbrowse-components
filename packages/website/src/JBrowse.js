import { readConfObject } from '@gmod/jbrowse-core/configuration'
import { App, theme } from '@gmod/jbrowse-core/ui'
import { toUrlSafeB64 } from '@gmod/jbrowse-core/util'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/core/styles'
import { observer } from 'mobx-react'
import { getSnapshot, onSnapshot } from 'mobx-state-tree'
import { StringParam, useQueryParam } from 'use-query-params'
import React, { useEffect } from 'react'

const MAX_SESSION_SIZE_IN_URL = 100000

// adapted from https://github.com/jashkenas/underscore/blob/5d8ab5e37c9724f6f1181c5f95d0020815e4cb77/underscore.js#L894-L925
function debounce(func, wait) {
  let timeout
  let result
  const later = (...args) => {
    timeout = null
    result = func(...args)
  }
  const debounced = (...args) => {
    if (timeout) {
      clearTimeout(timeout)
    }
    timeout = setTimeout(() => {
      return later(...args)
    }, wait)
    return result
  }
  debounced.cancel = () => {
    clearTimeout(timeout)
    timeout = null
  }
  return debounced
}

const JBrowse = observer(({ pluginManager }) => {
  const [, setSession] = useQueryParam('session', StringParam)

  const { rootModel } = pluginManager
  const { session, jbrowse, error } = rootModel || {}
  const useLocalStorage = jbrowse
    ? readConfObject(jbrowse.configuration, 'useLocalStorage')
    : false

  const useUpdateUrl = jbrowse
    ? readConfObject(jbrowse.configuration, 'useUrlSession')
    : false

  // Set session URL on first render only, before `onSnapshot` has fired
  useEffect(() => {
    if (useUpdateUrl) {
      const json = JSON.stringify(getSnapshot(session))
      const sess =
        json.length < MAX_SESSION_SIZE_IN_URL ? toUrlSafeB64(json) : undefined
      setSession(sess)
    }
  }, [session, setSession, useUpdateUrl])

  useEffect(() => {
    function updateUrl(snapshot) {
      const json = JSON.stringify(snapshot)
      const sess =
        json.length < MAX_SESSION_SIZE_IN_URL ? toUrlSafeB64(json) : undefined

      setSession(sess)
      if (rootModel && rootModel.session) {
        rootModel.jbrowse.updateSavedSession(snapshot)
      }
      if (useLocalStorage) {
        localStorage.setItem('jbrowse-web-session', JSON.stringify(snapshot))
      }
    }

    let disposer = () => {}
    if (session && useUpdateUrl) {
      const updater = debounce(updateUrl, 400)
      const snapshotDisposer = onSnapshot(session, updater)
      disposer = () => {
        snapshotDisposer()
        updater.cancel()
      }
    }
    return disposer
  }, [rootModel, setSession, useLocalStorage, useUpdateUrl, session])

  if (error) {
    throw new Error(error)
  }

  return <App session={rootModel.session} />
})

export default props => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <JBrowse {...props} />
    </ThemeProvider>
  )
}
