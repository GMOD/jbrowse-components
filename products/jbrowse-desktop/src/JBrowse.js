import { getConf } from '@jbrowse/core/configuration'
import { App, StartScreen, createJBrowseTheme } from '@jbrowse/core/ui'

import CssBaseline from '@material-ui/core/CssBaseline'

import { ThemeProvider } from '@material-ui/core/styles'

import { observer } from 'mobx-react'
import { onSnapshot } from 'mobx-state-tree'
import React, { useEffect, useState } from 'react'
import factoryReset from './factoryReset'

const debounceMs = 1000

// adapted from https://github.com/jashkenas/underscore/blob/5d8ab5e37c9724f6f1181c5f95d0020815e4cb77/underscore.js#L894-L925
function debounce(func, wait) {
  let timeout
  let result
  const later = (...args) => {
    timeout = null
    result = func(...args)
  }
  const debounced = (...args) => {
    if (timeout) {
      clearTimeout(timeout)
    }
    timeout = setTimeout(() => {
      return later(...args)
    }, wait)
    return result
  }
  debounced.cancel = () => {
    clearTimeout(timeout)
    timeout = null
  }
  return debounced
}

const JBrowse = observer(({ pluginManager }) => {
  const { electron = {} } = window
  const { desktopCapturer, ipcRenderer } = electron
  const [firstLoad, setFirstLoad] = useState(true)

  const { rootModel } = pluginManager
  const { session, jbrowse, error } = rootModel
  if (firstLoad && session) setFirstLoad(false)

  useEffect(() => {
    function sendIpcConfig(snapshot) {
      ipcRenderer.send('saveConfig', snapshot)
    }

    let disposer = () => {}
    if (jbrowse) {
      const updater = debounce(sendIpcConfig, debounceMs)
      const snapshotDisposer = onSnapshot(jbrowse, updater)
      disposer = () => {
        snapshotDisposer()
        updater.cancel()
      }
    }
    return disposer
  }, [jbrowse, ipcRenderer])

  useEffect(() => {
    async function createScreenshot(snapshot) {
      const sources = await desktopCapturer.getSources({
        types: ['window'],
        thumbnailSize: { width: 500, height: 500 },
      })
      const jbWindow = sources.find(source => source.name === 'JBrowse')
      const screenshot = jbWindow.thumbnail.toDataURL()
      ipcRenderer.send('saveSession', snapshot, screenshot)
    }

    let disposer = () => {}
    if (session) {
      const updater = debounce(createScreenshot, debounceMs)
      const snapshotDisposer = onSnapshot(session, updater)
      disposer = () => {
        snapshotDisposer()
        updater.cancel()
      }
    }
    return disposer
  }, [session, desktopCapturer, ipcRenderer])

  if (error) {
    throw new Error(error)
  }

  const theme = getConf(rootModel.jbrowse, 'theme')
  const { AssemblyManager } = pluginManager.getPlugin(
    'DataManagementPlugin',
  ).exports

  return (
    <ThemeProvider theme={createJBrowseTheme(theme)}>
      <CssBaseline />
      {rootModel.session ? (
        <>
          <App session={rootModel.session} />
          <AssemblyManager
            rootModel={rootModel}
            open={rootModel.isAssemblyEditing}
            onClose={() => {
              rootModel.setAssemblyEditing(false)
            }}
          />
        </>
      ) : (
        <StartScreen
          root={rootModel}
          bypass={firstLoad}
          onFactoryReset={factoryReset}
        />
      )}
    </ThemeProvider>
  )
})

export default JBrowse
