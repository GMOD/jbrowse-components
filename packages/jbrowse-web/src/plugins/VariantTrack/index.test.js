import { getSnapshot } from 'mobx-state-tree'
import MyPlugin from './index'
import { createTestEnv } from '../../JBrowse'

jest.mock('shortid', () => ({ generate: () => 'testid' }))

test('plugin in a stock JBrowse', async () => {
  const { pluginManager } = await createTestEnv()
  expect(() => pluginManager.addPlugin(new MyPlugin())).toThrow(
    /JBrowse already configured, cannot add plugins/,
  )

  const VariantTrack = pluginManager.getTrackType('VariantTrack')
  const config = VariantTrack.configSchema.create({ type: 'VariantTrack' })
  expect(getSnapshot(config)).toMatchSnapshot()
})
