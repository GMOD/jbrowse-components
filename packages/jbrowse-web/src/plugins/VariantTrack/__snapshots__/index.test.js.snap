// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin in a stock JBrowse 1`] = `
Object {
  "adapter": Object {
    "configId": "testid",
    "index": Object {
      "configId": "testid",
    },
    "type": "BamAdapter",
  },
  "configId": "testid",
  "renderers": Object {
    "PileupRenderer": Object {
      "configId": "testid",
      "type": "PileupRenderer",
    },
    "SvgFeatureRenderer": Object {
      "configId": "testid",
      "labels": Object {
        "configId": "testid",
      },
      "type": "SvgFeatureRenderer",
    },
    "configId": "testid",
  },
  "type": "VariantTrack",
}
`;
